[WIP] Draft application.

malukenho opened the pull request:
Starting to draw the basic features of the application.

Commit:
Update php version on composer file

Commit:
Fix controller name on router declaration

Commit:
Use ::class operator instead of strings.

Commit:
Don't use master branch on composer dependencies

Commit:
Fix license on top of files, turn docblock on comment.

Commit:
Declare all modules as dependencies

Commit:
Configure doctrine params on config files and update composer.json with dependencies

Commit:
Configuration for zfcuser

Commit:
Create client entity

Commit:
Add phpunit and phpcs to composer requirements

Commit:
Create Tickets Entity

Commit:
Create Consultant entity and rename property on client "dateRegister" -> "createdAt" to maintain pattern

Commit:
Create form for ticket

Commit:
Create controller for ticket with "create a ticket" action

Commit:
Register Ticket controller on module.config.php

Commit:
Register Ticket action router

Commit:
Create submit element on ticket form

Ocramius commented on composer.json:
```diff
@@ -13,11 +13,16 @@
     "homepage": "http://roave.com/",
     "require": {
-        "php": ">=5.4",
+        "php": ">=5.5",
```
`~5.6`

Ocramius commented on config/autoload/doctrine.global.php:
```diff
@@ -0,0 +1,21 @@
+                    'port' => '3306',
+                    'dbname' => 'tickets',
+                    'driverOptions' => [
+                        PDO::MYSQL_ATTR_INIT_COMMAND => "SET NAMES 'UTF8'"
```
You can simply use `'encoding' => 'utf8'`. Works since 5.3.8 IIRC

Ocramius commented on config/autoload/zfcuser.global.php:
```diff
@@ -0,0 +1,24 @@
+
+use Zend\Db\Adapter\Adapter;
+
+$settings = [
```
Aggregate into the following array

Ocramius commented on module/Application/config/module.config.php:
```diff
@@ -46,14 +73,15 @@
         'abstract_factories' => [
-            'Zend\Cache\Service\StorageCacheAbstractServiceFactory',
-            'Zend\Log\LoggerAbstractServiceFactory',
+            StorageCacheAbstractServiceFactory::class,
```
I would suggest killing both of these abstract factories.

Ocramius commented on module/Application/config/module.config.php:
```diff
@@ -67,4 +95,12 @@
+        'driver' => [
+            [
+                'paths' => [__DIR__ . '/../src/Application/Entity']
+            ]
```
Add trailing commas everywhere, please

Ocramius commented on module/Application/src/Application/Controller/TicketController.php:
```diff
@@ -0,0 +1,35 @@
+{
+    public function openAction()
+    {
+        $form = new Ticket();
```
Forms should always be built via the form element manager

Ocramius commented on module/Application/src/Application/Entity/Client.php:
```diff
@@ -0,0 +1,77 @@
+namespace Application\Entity;
+
+/**
+ * @Entity
```
Please use XML mappings if possible.

If not, keep in mind that DoctrineORMModule does not support this annotation syntax: you must import the annotation namespace:

``` php
use Doctrine\ORM\Mapping as ORM;

/**
 * @ORM\Entity
 */
```

Ocramius commented on module/Application/src/Application/Entity/Client.php:
```diff
@@ -0,0 +1,77 @@
+    /**
+     * @Column(type="datetime")
+     */
+    private $birthday;
```
Not needed

Ocramius commented on module/Application/src/Application/Entity/Client.php:
```diff
@@ -0,0 +1,77 @@
+    /**
+     * @Column(type="varchar", length="50", nullable="true")
+     */
+    private $phone;
```
Not needed: we can move these details into a `Contact` VO if necessary

Ocramius commented on module/Application/src/Application/Entity/Client.php:
```diff
@@ -0,0 +1,77 @@
+    /**
+     * @Column(name="mobile_phone", type="varchar", length="50", nullable="true")
+     */
+    private $mobilePhone;
```
Remove

Ocramius commented on module/Application/src/Application/Entity/Client.php:
```diff
@@ -0,0 +1,77 @@
+    /**
+     * @Column(type="varchar", length="255")
+     */
+    private $address;
```
Remove

Ocramius commented on module/Application/src/Application/Entity/Client.php:
```diff
@@ -0,0 +1,77 @@
+    /**
+     * @Column(type="text")
+     */
+    private $observation;
```
Not needed. Consider that this is not going to be a CRM (for now)

Ocramius commented on module/Application/src/Application/Entity/Client.php:
```diff
@@ -0,0 +1,77 @@
+    /**
+     * @Column(name="created_at", type="datetime")
+     */
+    private $createdAt;
```
Skip this: we are going to use CQS or CQRS+ES, so we don't need to track this sort of information, as it will emerge from logs

Ocramius commented on module/Application/src/Application/Entity/Consultant.php:
```diff
@@ -0,0 +1,69 @@
+namespace Application\Entity;
+
+/**
+ * @Entity
```
Same as above

Ocramius commented on module/Application/src/Application/Entity/Consultant.php:
```diff
@@ -0,0 +1,69 @@
+    /**
+     * @Column(type="varchar", length="255")
+     */
+    private $facebook;
```
Reduce contact details to the bare minimum: we will use a `Contact` VO (if needed) in future, instead.

Ocramius commented on module/Application/src/Application/Entity/Consultant.php:
```diff
@@ -0,0 +1,69 @@
+    /**
+     * @Column(type="boolean")
+     */
+    private $status;
```
What is `status`? Not explicative enough

Ocramius commented on module/Application/src/Application/Entity/Ticket.php:
```diff
@@ -0,0 +1,65 @@
+namespace Application\Entity;
+
+/**
+ * @Entity
```
Same as above

Ocramius commented on module/Application/src/Application/Entity/Client.php:
```diff
@@ -0,0 +1,77 @@
+     * @GeneratedValue
+     * @Column(type="integer", length="20")
+     */
+    private $id;
```
Let's use UUIDs for this: `@GeneratedValue(strategy="NONE")`

Ocramius commented on module/Application/src/Application/Entity/Consultant.php:
```diff
@@ -0,0 +1,69 @@
+{
+    /**
+     * @Id
+     * @GeneratedValue
```
Let's use UUIDs for this: `@GeneratedValue(strategy="NONE")`

Ocramius commented on module/Application/src/Application/Entity/Consultant.php:
```diff
@@ -0,0 +1,69 @@
+    /**
+     * @Column(name="created_at", type="datetime")
+     */
+    private $createdAt;
```
Same as above

Ocramius commented on module/Application/src/Application/Entity/Ticket.php:
```diff
@@ -0,0 +1,65 @@
+{
+    /**
+     * @Id
+     * @GeneratedValue
```
Let's use UUIDs for this: `@GeneratedValue(strategy="NONE")`

Ocramius commented on module/Application/src/Application/Entity/Ticket.php:
```diff
@@ -0,0 +1,65 @@
+    /**
+     * @Column(name="project_id", type="integer")
+     */
+    private $projectId;
```
Don't reference entities via ID: use an actual association instead

Ocramius commented on module/Application/src/Application/Form/Ticket.php:
```diff
@@ -0,0 +1,123 @@
+        ]);
+
+        $this->add([
+            'name' => 'created_at',
```
No need ;-)

Ocramius commented on module/Application/view/application/ticket/open.phtml:
```diff
@@ -0,0 +1,33 @@
+$form = $this->form;
+$form->setAttribute('action', $this->url('register-ticket'));
+
+echo $this->form()->openTag($form);
```
Unless this form really looks bad without all this additional markup, you can just use `<?= $this->form($form); ?>`

Commit:
Use doctrine `encode` config to set names to utf8

Commit:
Join array configuration on ZFCUser configuration

Commit:
Add trailing comma on doctrine config declaration

Commit:
Remove abstract_factories

Commit:
Remove unnecessary namespace declaration on module.config

Commit:
Use Form Element Manager to create a form object on ticket controller

Commit:
Remove fields from entity and form

Commit:
Remove old fields from generate form on view

Commit:
Rename status to active field

Commit:
Update php version for 5.6

Commit:
use Doctrine\ORM\Mapping namespace on entities

Commit:
Fix form declaration type key are defined on a wrong place and align composer require-dev dependencies version

Commit:
Travis CI configuration

Commit:
Create tests folder

Commit:
Get/Set on Ticket Entity

Commit:
Fill Entity with information

Commit:
Create TicketController#indexAction for display tickets and the template correspondent

Commit:
Ticket Command implementation

Ocramius commented on module/Application/src/Application/Command/Ticket/TicketCommand.php:
```diff
@@ -0,0 +1,36 @@
+use Doctrine\ORM\EntityManager;
+use Zend\Di\ServiceLocator;
+
+class TicketCommand
```
This is not a Command, it's a Command-handler

Ocramius commented on module/Application/src/Application/Command/Ticket/TicketCommand.php:
```diff
@@ -0,0 +1,36 @@
+    public function handleOpenNewTicket(OpenNewTicket $command)
+    {
+        $serviceLocator = new ServiceLocator();
+        $em = $serviceLocator->get(EntityManager::class)
```
Inject the EntityManager instead of using the ServiceLocator here

Ocramius commented on module/Application/src/Application/Controller/TicketController.php:
```diff
@@ -0,0 +1,58 @@
+    public function openAction()
+    {
+        $form = $this
+            ->getServiceLocator()
```
Inject instead of using the locator

Ocramius commented on module/Application/src/Application/Controller/TicketController.php:
```diff
@@ -0,0 +1,58 @@
+        ]);
+    }
+
+    public function registerAction()
```
action should probably reflect the command name

Ocramius commented on module/Application/src/Application/Controller/TicketController.php:
```diff
@@ -0,0 +1,58 @@
+        $ticket = new TicketEntity();
+        $ticket->fillEntity($params);
+
+        $ticketCommand = new TicketCommand();
```
The idea is kinda ok, except that you don't use the handler directly inside the `registerAction`

Ocramius commented on module/Application/src/Application/Controller/TicketController.php:
```diff
@@ -0,0 +1,58 @@
+        $ticket->fillEntity($params);
+
+        $ticketCommand = new TicketCommand();
+        $ticketCommand->handleOpenNewTicket(new OpenNewTicket($ticket));
```
The command should just be pushed down to a command bus: we can discuss that later in chat

Commit:
Turn TicketCommand on TicketHandler

Commit:
Inject dependencies on TicketController

Ocramius commented on config/autoload/zfcuser.global.php:
```diff
@@ -0,0 +1,22 @@
+    ],
+    'service_manager' => [
+        'aliases' => [
+            'zfcuser_zend_db_adapter' => Adapter::class,
```
We really shouldn't use `Zend\Db`

Ocramius commented on module/Application/src/Application/Command/Ticket/OpenNewTicket.php:
```diff
@@ -0,0 +1,36 @@
+{
+    private $ticketEntity;
+
+    public function __construct(Ticket $ticket)
```
I think this command should only accept a ticket ID, not the Ticket itself

Ocramius replied:
Also, you cannot really have a `OpenNewTicket` command with an existing ticket... This command should probably hold data required to _build_ a ticket, not a ticket directly.

malukenho replied:
@Ocramius I don't have a `Id` in the same way.

Ocramius replied:
I know: you just have the data to open a new ticket ;-)

Ocramius commented on module/Application/src/Application/Controller/TicketController.php:
```diff
@@ -0,0 +1,73 @@
+    {
+        $form = $this
+            ->formManager
+            ->get(Ticket::class);
```
I think we should use `create` here

Ocramius replied:
Note: `create` is suggested so that we avoid sharing the form across multiple actions. We may as well inject the form and make it `'shared' => false` instead

Commit:
Update doctrine config

Ocramius commented on module/Application/config/module.config.php:
```diff
@@ -67,4 +112,22 @@
+                'class' => AnnotationDriver::class,
+                'cache' => 'array',
+                'paths' => [
+                    __DIR__ . '/src/Application/Entity'
```
Are you sure about this path?

malukenho replied:
I try the `__DIR__ . '/../src/Application/Entity'` but the same happens.

Ocramius replied:
What does `./vendor/bin/doctrine-module orm:info` report?

What is the realpath here?

malukenho replied:
```
./vendor/bin/doctrine-module orm:info
Found 2 mapped entities:
[OK]   ZfcUser\Entity\User
[OK]   ZfcUserDoctrineORM\Entity\User
```

Commit:
Update composer.lock hash

Ocramius commented on module/Application/config/module.config.php:
```diff
@@ -67,4 +112,22 @@
+            ],
+        ],
+
+        'orm_default' => [
```
This key is nested incorrectly.

Commit:
Fix doctrine configuration

Commit:
Fix entities types definition

Commit:
Simplify TicketCommands by create abstraction

Commit:
Update information for ticket entity

Commit:
Create a ticket identifier

Commit:
Remove ticket command

Commit:
Handler to remove a ticket

Commit:
Refactor TicketHandler -> TicketCommandHandler to be more declarative

Commit:
Use TicketCommandHandler on import namespace

Commit:
Create remove ticket action

Commit:
Register router to remove a ticket

Commit:
List tickets on ticket home page and insert ticket link on menu top

Commit:
Fix phpcs on TicketIdentifier, only one blank line before namespace declaration

Commit:
Put entity manager as a dependency of TicketCommandHandler

Commit:
Remove old params from ticket controller

Commit:
Create corporation controller

Ocramius commented on composer.json:
```diff
@@ -13,11 +13,16 @@
+        "doctrine/doctrine-module":         "0.8.*",
         "zendframework/zendframework":      "2.3.*",
-        "zf-commons/zfc-user-doctrine-orm": "dev-master"
+        "zf-commons/zfc-user-doctrine-orm": "1.0.0",
```
Always let the patch version unbound: `1.0.*` ;-)

Ocramius commented on config/autoload/zfcuser.global.php:
```diff
@@ -0,0 +1,23 @@
+
+    'service_manager' => [
+        'aliases' => [
+            'zfcuser_zend_db_adapter' => Adapter::class,
```
Can probably die

Ocramius commented on module/Application/config/module.config.php:
```diff
@@ -67,4 +128,19 @@
+
+    'doctrine' => [
+        'driver' => [
+            'application_entity' => [
```
In a subsequent PR, we should consider renaming `Application`. Opened #8 for this.

Ocramius commented on module/Application/src/Application/Command/Ticket/OpenNewTicket.php:
```diff
@@ -0,0 +1,23 @@
+
+namespace Application\Command\Ticket;
+
+class OpenNewTicket extends TicketCommand
```
No properties? ;-) You need to start thinking about what composes a ticket's fundamentals ;-)

Ocramius commented on module/Application/src/Application/Command/Ticket/TicketCommand.php:
```diff
@@ -0,0 +1,47 @@
+
+use Application\Entity\Ticket;
+
+abstract class TicketCommand
```
I wouldn't use an "abstract command" here. Just code each command separately (no problem if there's duplication)

Ocramius commented on module/Application/src/Application/Command/Ticket/TicketCommand.php:
```diff
@@ -0,0 +1,47 @@
+    public function __construct($subject, $description, $importance, $projectId, $openedBy)
+    {
+        $this->ticketEntity = new Ticket();
+        $this->ticketEntity->setSubject($subject);
```
This command is altering the ticket. I feel like there is some misunderstanding here: maybe we should have a chat tomorrow, and go through the code by pair-programming a bit.

Ocramius replied:
In general, a `Command` does nothing: it just contains information, but it doesn't apply any action

Ocramius commented on module/Application/src/Application/Command/Ticket/TicketCommandHandler.php:
```diff
@@ -0,0 +1,52 @@
+        $this->entityManager = $entityManager;
+    }
+
+    public function handleOpenNewTicket(OpenNewTicket $command)
```
This should create a new ticket entity, assign properties and persist/flush

Ocramius commented on module/Application/src/Application/Command/Ticket/TicketCommandHandler.php:
```diff
@@ -0,0 +1,52 @@
+
+    public function handleOpenNewTicket(OpenNewTicket $command)
+    {
+        $this->entityManager->persist($command->getEntity());
```
This should create a new ticket entity, assign properties and persist/flush (instead of getting the already built entity from the command)

Ocramius commented on module/Application/src/Application/Command/Ticket/TicketCommandHandler.php:
```diff
@@ -0,0 +1,52 @@
+    {
+        $entity = $this->entityManager
+            ->getRepository(Ticket::class)
+            ->findOneBy(['id' => $command->getIdentifier()]);
```
`getTicketIdentifier` is more explicit here

Ocramius commented on module/Application/src/Application/Command/Ticket/TicketIdentifier.php:
```diff
@@ -0,0 +1,49 @@
+     */
+    public function __construct($identifier)
+    {
+        if (! is_numeric($identifier)) {
```
We should probably use UUIDs here ;-)

Ocramius replied:
I suggest https://github.com/ramsey/uuid unless @ramsey has better libraries to suggest/look at

ramsey replied:
Why would I have better libraries to suggest? Mine is the best! ;-)

Ocramius replied:
@ramsey oh, the humility ;-)

ramsey replied:
In all seriousness, I created it because I felt that all the others were lacking in different ways. I haven't seen others recently that generate UUIDs better, and the forthcoming 3.0 version will allow the ability to inject your own PRNGs, etc. into it.

Ocramius replied:
Awesome :) I guess we have a winner ;-)

Ocramius commented on module/Application/src/Application/Command/Ticket/TicketIdentifier.php:
```diff
@@ -0,0 +1,49 @@
+
+use InvalidArgumentException;
+
+class TicketIdentifier
```
Since we may use this in entities, consider providing also a `__toString` method (this is an ORM limitation)

Ocramius commented on module/Application/src/Application/Controller/TicketController.php:
```diff
@@ -0,0 +1,107 @@
+    {
+        $params = $this->params()->fromPost();
+
+        $this->commandTicket->handleOpenNewTicket(
```
We need to introduce the command bus here: we'll do it tomorrow if you're fine with that.

Ocramius commented on module/Application/src/Application/Entity/Client.php:
```diff
@@ -0,0 +1,59 @@
+    /**
+     * @ORM\Id
+     * @ORM\GeneratedValue
+     * @ORM\Column(type="integer", length=20)
```
Avoid using `GeneratedValue`: let's just use UUIDs

ramsey replied:
FYI, someone contributed a Doctrine UuidType to rhumsaa/uuid that you might find helpful: https://github.com/ramsey/uuid/blob/master/src/Doctrine/UuidType.php

Ocramius replied:
Yeah, that's what I was thinking of, though we'll have to extend it to have specific identifier classes

Ocramius replied:
For now it should be enough to just store the UUID as string and deal with it in the methods

Ocramius commented on module/Application/src/Application/Entity/Consultant.php:
```diff
@@ -0,0 +1,47 @@
+    /**
+     * @ORM\Id
+     * @ORM\GeneratedValue
+     * @ORM\Column(name="id", type="integer")
```
Avoid using `GeneratedValue`: let's just use UUIDs

Ocramius commented on module/Application/src/Application/Entity/Ticket.php:
```diff
@@ -0,0 +1,201 @@
+    /**
+     * @ORM\Id
+     * @ORM\GeneratedValue
+     * @ORM\Column(name="id", type="integer")
```
Avoid using `GeneratedValue`: let's just use UUIDs

Ocramius commented on module/Application/src/Application/Entity/Ticket.php:
```diff
@@ -0,0 +1,201 @@
+        $this->active = $active;
+    }
+
+    public function fillEntity(array $data)
```
This method should be killed. Most setters/getters are also not needed. Remove them unless they are strictly necessary.

Commit:
Remove TicketCommand abstract class

Commit:
Put wildcard on version

Commit:
Remove Application\Entity\Ticket#fillEntity method

Commit:
Remove configuration to Zend\Db from zfcuser

Ocramius committed:
Applying inline notes for PR #6 review

Commit:
Remove application router

Commit:
Merge pull request #1 from Roave/hotfix/#6-review

@Ocramius review

Commit:
Use enum implementation to Ticket form

Commit:
Use child_routes to describe ticket sub routes

Commit:
Simplify ticket identifier and be more specific on ticketController

Commit:
Fix redirect router correctly on post open ticket

Commit:
Register commandBus on service manager and pass specific repository to ticketController

Commit:
Update composer.lock

Commit:
Remove TicketIdentifier

Commit:
Fix OpenNewTicket CS, module.config urls and urls on views

Commit:
Register View router

Commit:
Prepare tables to use UUIDs

Commit:
Send form to open a new ticket to the same router

Commit:
Add a view ticket view

Commit:
Create filter to ticket form

Commit:
Write a commandBus for tickets command and fix another command

Commit:
Fix entity ticket

Commit:
Create event for TicketRemoved

Commit:
Fix dependency on ticketCommandHandler on module.config.php

Commit:
Remove unnecessary configuration from module config, fix code style on
ticket controller and starting to create Consultant Controller.

Commit:
- Fix filter for ticket form and redirect to "view" ticket router correctly.
- Create template for view ticket route, showing basic information about the tickets.
- Modify exception from \\Exception to \\BadMethodCallException, on Filter\Ticket#setInputFilter(), because it's make more sense.
- Use camel case syntax on validators names ('not_empty' --> 'NotEmpty')
- Put view link on ticket list
- Modify imported namespace (TicketFilter --> TicketFormFilter) on TicketController

Commit:
Possibility of update a ticket information

Commit:
Create commands to SOLVE and CLOSE a ticket, also fix pattern on param URL

Commit:
Implement comment on tickets

Commit:
Display comments on view ticket page

Commit:
Implement re-open ticket command

Commit:
Fix php cs

Commit:
Remove TicketCommandHandler class and separate handlers on small classes, create a abstract class AbstractCommandHandler to abstract handlers and a new namespace to put handlers \Application\Command\Ticket\Handler;

CommandBus#handler is renamed to CommandBus#push and modify verification id a command can be handled by using a method on handlers named AbstractCommandHandler#canHandler().

All commands need to implements the Command interface now.

Follow is the new way:

  +----------------------+
  |      CommandBus      |
  +----------------------+
  |  + push(Command)     |
  +----------+-----------+
             |
             |___________________\  Can any handler registered
                                 /      handler this guy?
                                                |
                                                |
        Handler#handler(Command) /______________|
           return a event        \

Ocramius committed:
Removing example command bus

Ocramius committed:
s/handler/handle

Ocramius committed:
s/canHandler/canHandle

Ocramius committed:
Implementing `CommandHandlerInterface` instead of extending an abstract type

Ocramius committed:
Removing base command handler (not needed)

Ocramius committed:
Making all command handlers `final`

Ocramius committed:
Command bus return value should be `void`, and handlers type should be `CommandHandlerInterface`

Ocramius committed:
Removing useless IDE hint (hinting on property instead)

Ocramius committed:
Better type safety inside the command bus

Ocramius committed:
Optimized imports

Commit:
Register a listener to verify if user is logged

danizord commented on README.md:
```diff
@@ -1,2 +1,14 @@
+ 1. http request hits the application (application layer)
+ 2. zf2 runs the respective controller (application layer)
+ 3. controller validates data (application layer)
+ 4. controller creates command (domain layer)
```
app layer?

danizord commented on module/Application/config/module.config.php:
```diff
@@ -45,15 +172,57 @@
-            'Zend\Log\LoggerAbstractServiceFactory',
-        ]
+        'factories' => [
+            CommandBus::class => function ($em) {
```
Closures makes confing unserializable. This factory should be moved to a class

gabrielsch commented on module/Application/src/Application/Entity/Ticket.php:
```diff
@@ -0,0 +1,162 @@
+     */
+    private $solved;
+
+    public function __construct()
```
you could ensure object consistence by requiring some arguments in constructor, for example, can a ticket exist without subject or description?

malukenho replied:
@gabrielsch for this the form has validations and required fields, nops?

danizord commented on module/Application/config/module.config.php:
```diff
@@ -45,15 +172,57 @@
+        ],
+
+        'factories' => [
+            TicketController::class => function ($em) {
```
Same as above

danizord commented on module/Application/src/Application/Command/AbstractCommandHandler.php:
```diff
@@ -0,0 +1,38 @@
+
+    abstract public function handler(Command $command);
+
+    abstract public function canHandler(Command $command);
```
canHandle?

danizord commented on module/Application/src/Application/Command/AbstractCommandHandler.php:
```diff
@@ -0,0 +1,38 @@
+     */
+    protected $entityManager;
+
+    public function __construct(EntityManager $entityManager)
```
Should probably typehint ObjectManager

danizord commented on module/Application/src/Application/Command/Ticket/CloseTicket.php:
```diff
@@ -0,0 +1,45 @@
+    /**
+     * @param string $uuid
+     */
+    public function __construct($uuid)
```
Should probably work with Uuid instance

danizord commented on module/Application/src/Application/Command/Ticket/CommentOnTicket.php:
```diff
@@ -0,0 +1,61 @@
+
+    public function __construct(Ticket $ticket, $comment)
+    {
+        $this->uuid    = Uuid::uuid1();
```
I think it'd be better to generate UUID outside (in controller)

danizord commented on module/Application/src/Application/Command/Ticket/CommentOnTicket.php:
```diff
@@ -0,0 +1,61 @@
+    public function __construct(Ticket $ticket, $comment)
+    {
+        $this->uuid    = Uuid::uuid1();
+        $this->ticket  = $ticket;
```
Should not be TicketId instead?

danizord commented on module/Application/src/Application/Command/Ticket/Handler/CloseTicket.php:
```diff
@@ -0,0 +1,45 @@
+    {
+        $entity = $this->entityManager
+            ->getRepository(Ticket::class)
+            ->findOneBy(['id' => $command->getTicketIdentifier()]);
```
Could be `find()`

danizord commented on module/Application/src/Application/Command/Ticket/Handler/CloseTicket.php:
```diff
@@ -0,0 +1,45 @@
+            ->getRepository(Ticket::class)
+            ->findOneBy(['id' => $command->getTicketIdentifier()]);
+
+        $entity->markAsClosed();
```
Could be `close()`

danizord commented on module/Application/src/Application/Command/Ticket/Handler/CloseTicket.php:
```diff
@@ -0,0 +1,45 @@
+        $entity->markAsClosed();
+        $this->entityManager->flush();
+
+        return new TicketWasClosed($command->getTicketIdentifier());
```
I think you should inject the event dispatcher and call it to dispatch events (so that you can dispatch more than 1 event)

danizord commented on module/Application/src/Application/Command/Ticket/Handler/CommentOnTicket.php:
```diff
@@ -0,0 +1,50 @@
+{
+    public function handler(Command $command)
+    {
+        $comment = new Comment();
```
Should be `new Comment($command->getComment(), $command->getTicket());`

danizord commented on module/Application/src/Application/Command/Ticket/Handler/OpenNewTicket.php:
```diff
@@ -0,0 +1,51 @@
+{
+    public function handler(Command $command)
+    {
+        $ticket = new Ticket();
```
Same as above

danizord commented on module/Application/src/Application/Command/Ticket/Handler/RemoveTicket.php:
```diff
@@ -0,0 +1,45 @@
+            ->getRepository(Ticket::class)
+            ->findOneBy(['id' => $command->getTicketIdentifier()]);
+
+        $this->entityManager->remove($entity);
```
What if the ticket was not found?

danizord commented on module/Application/src/Application/Command/Ticket/Handler/ReopenTicket.php:
```diff
@@ -0,0 +1,45 @@
+            ->getRepository(Ticket::class)
+            ->findOneBy(['id' => $command->getTicketIdentifier()]);
+
+        $entity->markAsOpened();
```
reopen()

danizord commented on module/Application/src/Application/Command/Ticket/Handler/SolveTicket.php:
```diff
@@ -0,0 +1,45 @@
+            ->getRepository(Ticket::class)
+            ->findOneBy(['id' => $command->getTicketIdentifier()]);
+
+        $entity->markAsSolved();
```
solve()

danizord commented on module/Application/src/Application/Controller/TicketController.php:
```diff
@@ -0,0 +1,208 @@
+        CommandBus $commandHandler,
+        Ticket $ticketForm,
+        Comment $commentForm,
+        $ticketRepository,
```
Missing typehint

danizord commented on module/Application/src/Application/Controller/TicketController.php:
```diff
@@ -0,0 +1,208 @@
+        Ticket $ticketForm,
+        Comment $commentForm,
+        $ticketRepository,
+        $commentRepository
```
Missing typehint

danizord commented on module/Application/src/Application/Listener/Authentication.php:
```diff
@@ -0,0 +1,60 @@
+     */
+    public function onRoute(MvcEvent $e)
+    {
+        $routerMatch = $e->getRouteMatch();
```
You must bypass if $routeMatch === null (404)

Commit:
Merge pull request #2 from Roave/review/#6-second-review-round

Review - #6 second review round

Commit:
Use bower to manage packages, create a `.bowerrc` file to install libraries on `public/assets` path. Remove od stuff and link the new one on base layout. Move own asssets to a folder named "custom".

Commit:
Use a factory to create the CommandBus instead of a Closure.

Commit:
s/domain/application

Xerkus commented:
this PR missing unit tests

Xerkus commented on module/Application/src/Application/Service/TicketCommandBus.php:
```diff
@@ -0,0 +1,47 @@
+ * This software consists of voluntary contributions made by many individuals
+ * and is licensed under the MIT license.
+ */
+namespace Application\Service;
```
Common convention for factories is to add `Factory` to namespace or classname

Xerkus commented:
i suggest to merge this PR to develop branch and continue development from there using smaller PRs.

Commit:
s/service/factory

Commit:
Move Ticket Controller Closure Factory to a class

Commit:
Pass Uuid from command to the Ticket Entity

Commit:
Remove return from Handler\<HandleObject>#handle(), the commandBus return nothing, so have no sense to make handler return anything.

Commit:
TicketController modification

- TypeCast on TicketController#__construct() for EntityRepository (last tho variables)
- Use the Uuid generated on Command to redirect to a router
- Use EntityRepository#find() instead of EntityRepository#findBy(<id: value>)

Commit:
s/findOneById/find

Commit:
Command Handler improve

 - Rename local variable $entity -> $ticket to increase readability
 - Use EntityRepository#find() instead of EntityRepository#findOneBy()
 - Rename methods from Ticket Entity
    - Ticket#markAsSolved   -> Ticket#solve
    - Ticket#markAsOpened -> Ticket#reopen
    - Ticket#markAsClosed  -> Ticket#close

Commit:
Fix docBlock no Comment Entity

Commit:
s/Application/Support/ namespace and close #8
